Opt for .cloned() over .map(|foo| foo.clone()) etc.

Ryman opened the pull request:
This overlaps with #22276 (I left make check running overnight) but covers a number of additional cases and has a few rewrites where the clones are not even necessary.

This also implements `RandomAccessIterator` for `iter::Cloned`

cc @steveklabnik, you may want to glance at this before #22281 gets the bors treatment

rust-highfive commented:
r? @nikomatsakis

(rust_highfive has picked a reviewer for you, use r? to override)

nikomatsakis commented:
Looks good to me!

nikomatsakis commented:
@bors r+ b042d10

bors commented:
:scream_cat: You have the wrong number! Please try again with `43f0fbf`.

nikomatsakis commented:
@bors r+ 43f0fbf

nikomatsakis commented:
@Ryman does this cover *all* the cases of #22276?

Ryman commented:
@nikomatsakis It missed `.map(|&foo| foo)`, which I just realised after @Gankro's comment on #22276. I could add that here, or as a new PR depending on preference? (Although I think @Gankro covered the ones in libcollections for that pattern)

nikomatsakis commented:
@bors r-

nikomatsakis commented:
I'm canceling the r+ until the other PRs land so we can sort this out in a more orderly fashion.

steveklabnik commented:
@bors: r-

Ryman commented:
Rebased on master and added the remaining cases.

r? @nikomatsakis

Gankro commented on src/librustc/middle/const_eval.rs:
```diff
@@ -600,7 +600,7 @@ pub fn lit_to_const(lit: &ast::Lit) -> const_val {
         ast::LitStr(ref s, _) => const_str((*s).clone()),
         ast::LitBinary(ref data) => {
-            const_binary(Rc::new(data.iter().map(|x| *x).collect()))
+            const_binary(data.clone())
```
This has slightly different semantics from the above code. I'm not clear what was intended. Naively this should be a strict improvement? (just bumps a refcount instead of creating a new Rc and new Vec)

librustc *is* crazy, though...

Ryman replied:
Yeah, assumed safe rust usage which makes this equivalent? Tests passed for a full `make check` so it should be fine?

Gankro replied:
If an Rc happens to be unique, you can actually move the data out. So there is actually a semantic change in safe Rust to make a new disjoint Rc. 

I don't think rustc ever actually does this, though.

Ryman replied:
Isn't that just CoW? Or has it changed recently? Wouldn't the underlying Vec be the same and whatever code wants to act on a mutation would Cow or clone the inside? It's not a RefCell so doing anything strange to the underlying data would be pretty dodgy? (Am I overlooking something really terrible?)

Other than this concern, rebase is done :)

Gankro replied:
Note to future: Talked to eddyb, this is just a case of gradual code refactoring making previously sane code insane.

Gankro commented on src/librustc_back/rpath.rs:
```diff
@@ -40,10 +40,7 @@ pub fn get_rpath_flags<F, G>(config: RPathConfig<F, G>) -> Vec<String> where
-        l.map(|p| p.clone())
-    }).collect::<Vec<_>>();
-
+    let libs = libs.into_iter().filter_map(|(_, l)| l).collect::<Vec<_>>();
```
fwiw `let libs: Vec<_> = ...` is more en-vogue right now. 

Also what happened here? The clone is just totally gone? I'm guessing the clone was totally redundant since this is a by-value iter?

Ryman replied:
Yeah, redundant it seems, if I remember, `cloned` wasn't implemented for `l` due to being an Option with an owned T. Is it a requirement to change the typehint in this PR?

Gankro replied:
I wouldn't block this on it, but since you're changing the line (and there's some merge conflict), it would be nice to change while you're in there.

Gankro commented on src/librustc_resolve/lib.rs:
```diff
@@ -1921,18 +1921,15 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
                                 -> Option<Rc<Module>> {
-            module.external_module_children.borrow()
-                                            .get(&needle).cloned()
-                                            .map(|_| module.clone())
```
heh, am I reading this right? It was basically double-cloning the data? Bulk-refactors are a hell of a drug. :)

Ryman replied:
Not quite, so I rewrote it to hopefully be clearer! It was cloning the `get` result and then just mapping it to an unrelated clone.

Gankro commented on src/libstd/collections/hash/set.rs:
```diff
@@ -292,21 +292,21 @@ impl<T, S, H> HashSet<T, S>
-    /// let a: HashSet<int> = [1, 2, 3].iter().map(|&x| x).collect();
-    /// let b: HashSet<int> = [4, 2, 3, 4].iter().map(|&x| x).collect();
+    /// let a: HashSet<int> = [1, 2, 3].iter().cloned().collect();
+    /// let b: HashSet<int> = [4, 2, 3, 4].iter().cloned().collect();
```
This file is all a dupe of https://github.com/rust-lang/rust/pull/22242 as far as I can tell.

Ryman replied:
Ah, filtered out `src/libcollections/*` but this is in std, I'll do a quick rebase on master since that one's merged now.

Gankro commented:
Wow, this is way more principled than a lot of bulk refactors we get! *Actually reflecting on if the surrounding code makes sense*?? :+1: 

Just a few minor issues.

Gankro commented:
@bors r+ a11e900

bors commented:
:hourglass: Testing commit a11e900 with merge 9c9711b...

bors commented:
:broken_heart: Test failed - [auto-win-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-32-nopt-t/builds/3299)

Ryman commented:
Windows failed with: 
```
libstd\sys\windows\thread_local.rs:247:42: 247:51 error: the trait `core::clone::Clone` is not implemented for the type `unsafe extern "C" fn(*mut u8)` [E0277]
libstd\sys\windows\thread_local.rs:247                 (*DTORS).iter().cloned().collect()
```

@Gankro re r? Rebased and reverted the above in last added commit as I don't think I can implement Clone for that? 

Related: are we ever going to have Copy implements Clone or was that written off for some reason?

Gankro commented:
Something Something Coherence makes the blanket impl hard to do, I think.

Gankro commented:
Can you squash the revert into whatever commit added it?

Ryman commented:
@Gankro done.

Gankro commented:
@bors r+ 20db2c

Gankro commented:
Thanks a ton!

Ryman commented:
@Gankro rebased as it seems to have dropped off bors's radar according to the build queue.

Gankro commented:
@bors r+ d2f54e

bors commented:
:hourglass: Testing commit d2f54e6 with merge 3d01ccc...

bors commented:
:broken_heart: Test failed - [auto-win-64-opt](http://buildbot.rust-lang.org/builders/auto-win-64-opt/builds/2173)

alexcrichton commented:
@bors: retry

bors commented:
:hourglass: Testing commit d2f54e6 with merge dfaf453...

bors commented:
:broken_heart: Test failed - [auto-win-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-32-nopt-t/builds/3347)

Gankro commented:
@bors retry
